runtime/gc_modron_standard/ReadBarrierVerifier.cpp

```diff
@@ -0,0 +1,292 @@
+	omrobjectptr_t object = convertPointerFromToken(objectToken);
+
+	GC_SlotObject slotObject(env->getOmrVM(), srcAddress);
+
+	if ((shadowHeapTop > (uintptr_t)object && shadowHeapBase <= (uintptr_t)object)) {
+		healedAddress = heapBase + ((uintptr_t)object - shadowHeapBase);
+
+		GC_SlotObject slotObject(env->getOmrVM(), srcAddress);
+		slotObject.atomicWriteReferenceToSlot(object, (omrobjectptr_t)healedAddress);
+	}
+	return true;
+}
+
+void
+MM_ReadBarrierVerifier::poisonSlot(MM_GCExtensionsBase *extensions, fomrobject_t *slot)
+{
+	uintptr_t heapBase = (uintptr_t)extensions->heap->getHeapBase();
+	uintptr_t heapTop = (uintptr_t)extensions->heap->getHeapTop();
+	uintptr_t shadowHeapBase = (uintptr_t)extensions->shadowHeapBase;
+
+	uintptr_t poisonedAddress;
+
+	uintptr_t referenceFromSlot = (uintptr_t)*slot;
+	if (referenceFromSlot < heapTop && referenceFromSlot >= heapBase) {
+		poisonedAddress = shadowHeapBase + (referenceFromSlot - heapBase);
+		*slot = (fomrobject_t)poisonedAddress;
+	} else {
+
+	}
+}
+
+void
+MM_ReadBarrierVerifier::poisonJniWeakReferenceSlots(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true, true);
+	scanner.scanJNIWeakGlobalReferences(env);
+}
+
+void
+MM_ReadBarrierVerifier::poisonMonitorReferenceSlots(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true, true);
+	scanner.scanMonitorReferences(env);
+}
+
+void
+MM_ReadBarrierVerifier::poisonStaticClassSlots(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true, true);
+	scanner.scanClassStaticSlots(env);
+}
+
+void
+MM_ReadBarrierVerifier::poisonConstantPoolObjects(MM_EnvironmentBase *env)
+{
+	MM_RootScannerReadBarrierVerifier scanner(env, true, true);
+	scanner.scanConstantPoolObjectSlots(env);
+}
+
+void
+MM_ReadBarrierVerifier::poisonSlots(MM_EnvironmentBase *env)
+{
+	MM_GCExtensionsBase *extensions = env->getExtensions();
+	if (1 == extensions->fvtest_enableJNIGlobalWeakReadBarrierVerification) {
+		poisonJniWeakReferenceSlots(env);
+	}
+	if (1 == extensions->fvtest_enableMonitorObjectsReadBarrierVerification) {
+		poisonMonitorReferenceSlots(env);
+	}
+	if (1 == extensions->fvtest_enableClassStaticsReadBarrierVerification) {
+		poisonStaticClassSlots(env);
+		poisonConstantPoolObjects(env);
+	}
+}
+
+void
+MM_ReadBarrierVerifier::healSlot(MM_GCExtensionsBase *extensions, j9object_t *srcAddress)
+{
+	uintptr_t heapBase = (uintptr_t)extensions->heap->getHeapBase();
+	uintptr_t shadowHeapBase = (uintptr_t)extensions->shadowHeapBase;
+	uintptr_t shadowHeapTop = (uintptr_t)extensions->shadowHeapTop;
+	uintptr_t healedAddress;
+	j9object_t object = *srcAddress;
+
+	if (shadowHeapTop > (uintptr_t)object && (shadowHeapBase <= (uintptr_t)object)) {
+		healedAddress = heapBase + ((uintptr_t)object - shadowHeapBase);
+
+		// TODO: instead of atomicWriteReference, write back to srcAddress. But maybe we might need it to be atomic
```

Comment:
let's play safe and use atomic update here. I belive that there could be a race where another thread can write a different value between the moment this thread read a poissoned value and the moment if wants to update with heal value. so, this thread should fail on update if the field has been modified meanwhile.

update two other spots too